Open documents in new tab

starofrainnight opened the pull request:
I think the title should be named as "Open documents in new tab when another ReText starts from console".

This PR just want to help complete #148 .

Tested under : 

1. python3.4 on win10 x64 
2. python3.5.1 on ubuntu16.04 x64

mitya57 commented on retext.py:
```diff
@@ -54,6 +57,39 @@ def main():
 	if hasattr(app, 'setDesktopFileName'): # available since Qt 5.7
 		app.setDesktopFileName('me.mitya57.ReText.desktop')
+
+	singleApp = SingleApplication("D1278E7822F011E6802800F1F38F93EF")
```
What does this number mean? :)

starofrainnight replied:
The number just an unique ID those all ReText shared, so that the new starting ReText will use to check if there already an ReText existed. It could be any text, but you must ensure it's unique value in system using by QSharedMemory (The key value of QSahredMemory).

The value in source code is generated by python code on my system:

`
import uuid
uuid.uuid1().hex.upper()
`

mitya57 replied:
How about just using "ReText" as the string?

starofrainnight replied:
OK, It will be changed to "ReText", I just concentrated upon unique value at the beginning.
Later it will be changed to something like "ReText-username" for multi-user that @Griffon26 referred.

mitya57 commented:
@Griffon26 You have more experience with sockets and multiprocessing, can I please ask you to review this?

Griffon26 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+		self._localSockets = {}
+
+		# Fix for *nix:
+		# http://stackoverflow.com/questions/5006547/qt-best-practice-for-a-single-instance-app-protection
```
Keep the links, but add an explanation of the fix here, because understanding the code should not depend on a web site.

starofrainnight replied:
OK, I will add an explanation here until I sorted them out later.

mitya57 commented on retext.py:
```diff
@@ -54,6 +57,39 @@ def main():
+	singleApp = SingleApplication("D1278E7822F011E6802800F1F38F93EF")
+	singleApp.start()
+	if singleApp.mode == singleApp.Client:
+		# When we in client mode, we just send the arguments to existed
```
When we **are** in client mode

mitya57 commented on retext.py:
```diff
@@ -54,6 +57,39 @@ def main():
+	singleApp.start()
+	if singleApp.mode == singleApp.Client:
+		# When we in client mode, we just send the arguments to existed
+		# appliaction and exit
```
to the exist**ing** appli**ca**tion instance

mitya57 commented on retext.py:
```diff
@@ -54,6 +57,39 @@ def main():
+		message = json.dumps(list(map(canonicalize, sys.argv[1:])))
+		singleApp.sendMessage(message.encode('utf-8'))
+		
+		# Exit the application immediately after send message 
```
after send**ing** the message

mitya57 commented on retext.py:
```diff
@@ -72,6 +108,8 @@ def main():
 		app.setStyleSheet(QTextStream(sheetfile).readAll())
 		sheetfile.close()
 	window = ReTextWindow()
+	# Mark this window so that we could find it while we received message. 
```
wh**en** we received the message

mitya57 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+# This file is part of ReText
+# Copyright: 2012-2015 Dmitry Shachnev
```
Please put your copyright here :)

mitya57 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+
+import functools
+import struct
+from PyQt5.QtCore import pyqtSignal
```
Please group the imports, i.e.:

    from PyQt5.QtCore import pyqtSignal, QObject, QSharedMemory, QSystemSemaphore

mitya57 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+		# Ensure we run only one application
+		isAnotherRunning = False
+		if not self._sharedMemory.isAttached():
+			# If share memory is attachable, that means there have another
```
share**d** memory
there **exists** another running application

mitya57 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+			finally:
+				self._systemSemaphore.release()
+
+			# If there something happened during create procedure lead us can't
```
Sorry, I can't parse this sentence…

If <s>there</s> something (what? maybe an example?) happened during the creation procedure and we can't create one (what?), we should detach that share**d** memory.

starofrainnight replied:
I can't answer these questions yet, because the whole single application framework is converted from some Qt/C++ principle code snippets that I still not fully understand. I just combine them together, make it seems works and try my best to comment them.

Why I comments here with "something" just I really don't know what will make this procedure failed.

They related to different system parts, and at the time those code snippets don't have much details about why they do that.

starofrainnight replied:
Maybe we don't need to comment here to tell what's going on, because we must do the clear job if the function failed to create, right?

I means, no matter what it happening, maybe system error or else, it must be clear after failed, so, we clear it.

mitya57 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+				finally:
+					self._systemSemaphore.release();
+
+		# Fine, now we could take different action on different situation.
```
This comment is not informative at all — if you want to comment here, describe *which* action we can take.

mitya57 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+					self._systemSemaphore.release();
+
+		# Fine, now we could take different action on different situation.
+		if not isAnotherRunning:
```
I prefer `if…else` blocks rather than `if not…else`.

mitya57 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+
+	def sendMessage(self, message):
+		# Only accept bytes message
+		assert(type(message) == bytes)
```
Was this assert just a debug, or really needed?
Also, `isinstance(message, bytes)` looks a bit more nice.

starofrainnight replied:
It's really needed.
I will change to `isinstance()` and raise TypeError if it failed.

starofrainnight replied:
It's really needed.
I will change `isinstance()`  and raise TypeError() if failed

Griffon26 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+		QObject.__init__(self, parent)
+		self._name = name
+		self._mode = self.Server
+		self._sharedMemory = QSharedMemory(self._name)
```
Each user on the system should be able to run this application, but currently the code forces one instance per system.
It would be easy enough to include the username here, but that would still make it very easy for a malicious user to perform a denial of service attack on others. Can you think of a way around this?

Griffon26 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+			self._server = QLocalServer(self)
+			self._server.newConnection.connect(self._onServerNewConnection)
+
+			if not self._server.listen(self._name):
```
I think we need QLocalServer::UserAccessOption here to avoid one user forcing another user's ReText to open a file.
Just like with the semaphore/shared memory we will also need a way to avoid denial of service. Is there a portable way to use a user-specific namespace, such as a file in the user's homedir?

Griffon26 commented on ReText/singleapplication.py:
```diff
@@ -0,0 +1,167 @@
+from PyQt5.QtNetwork import QLocalServer
+from PyQt5.QtNetwork import QLocalSocket
+
+class SingleApplication(QObject):
```
Given the complexity of solving this problem in a portable way and the risk of introducing problems that opens the application up to interference from other users, I'm wondering if there isn't some existing proven code that we can just reuse. Have you looked at alternatives and ruled them out?

starofrainnight commented:
WOW, that's so many questions and so many changes I needs to do ...
This PR is so rough to just make the function works.
Actually I'm not **very** clear about the details how it works on each platform, I trying to figure them out.
That need some time before I reply all your questions (I have to take more research and clear my mind), hope I have time to done all by near two weeks. 

First thing, I will try to fix my grammar errors, thanks your patient @mitya57. Seems that's the easier things I could do now.

mitya57 commented:
There is a [single application](http://code.qt.io/cgit/qt-solutions/qt-solutions.git/tree/qtsingleapplication) solution from Qt developers themselves. It relies on lock files so the application will have one instance per user, not per system. Also it uses QLocalSocket for messages exchange.

It was written before [QLockFile](http://doc.qt.io/qt-5/qlockfile.html) became part of Qt. Now that it's available, we can use it to simplify the task of porting that code to Python.

starofrainnight commented:
@mitya57 

I had go through that solution from Qt developers before this PR, it's seems too complicated to introduce a dozen classes into this project, so I left it for another. At that time, I don't know QLockFile has became part of Qt yet. Until you just refer to it.

@mitya57 @Griffon26 
I found a python single application implementation from tendo module, seems that if we introduce this module will solve our single application problem. It also relies on lock files. 

https://github.com/pycontribs/tendo/blob/master/tendo/singleton.py

There have two ways to rewrote this PR:

1. Rewrote by QLockFile
2. Introduce tendo module dependence and rewrote by it's SingleInstance class (I think this is more easy to implement)

Which one do you prefer?

Griffon26 commented:
Be sure to check the license of the tendo code. To me it's not clear if it falls under the python license and if so, which version. Not all python license versions are compatible with the GPL ReText is under.

mitya57 commented:
There are discussions in Qt upstream about including QSingleApplication in qtbase: http://lists.qt-project.org/pipermail/development/2016-June/026228.html

Let's see where it goes.

starofrainnight commented:
Yep, the license of the tendo code is not compatible with the GPL, so this way is down.
I trying reimplement by QLockFile.

mitya57 commented:
I think the security issue mentioned by @Griffon26 in https://github.com/retext-project/retext/pull/215#discussion_r66705329 is still there: the server is available for all users, and one can perform a DoS attack on it which can lead to data loss.

Let's maybe wait for QSingleApplication to become reality instead.

Don't take me wrong, but I started ReText as a simple and lightweight editor, and I didn't really expect it to get lots of code with potential security issues written not by me…

starofrainnight commented:
@mitya57 

You are right, the security issue still there, I still looking into it. But I have too busy to take that investigation these days. 

OK, lets wait for QSingleApplication to become reality instead.

I understand your serious attitude to the code, actually I like the way you are.
